ownCloud/Client/Actions/Actions+Extensions/DuplicateAction.swift

```diff
@@ -34,37 +37,126 @@ class DuplicateAction : Action {
+		let pattern = " \(localizedCopy)[ ]?[0-9]{0,}$"
+		let searchName = baseName.replacingOccurrences(for: pattern)
 
-				if let itemFileExtension = item.fileExtension, let baseName = item.baseName {
-					var fileExtension = itemFileExtension
+		query = OCQuery(forPath: rootItemPath)
+		query?.delegate = self
+		query?.queryItem = item
 
-					if fileExtension != "" {
-						fileExtension = ".\(fileExtension)"
-					}
+		let filterHandler: OCQueryFilterHandler = { (_, _, item) -> Bool in
+			if let itemName = item?.baseName {
+				return itemName.hasPrefix(searchName)
+			}
+			return false
+		}
+		query?.addFilter(OCQueryFilter.init(handler: filterHandler), withIdentifier: "rename-file")
 
-					name = "\(baseName) copy\(fileExtension)"
+		core.start(query!)
+	}
+
+	// MARK: - Query handling
+	func query(_ query: OCQuery, failedWithError error: Error) {
+		completed(with: NSError(ocError: .itemNotFound))
+	}
+
+	func queryHasChangesAvailable(_ query: OCQuery) {
+		query.requestChangeSet(withFlags: OCQueryChangeSetRequestFlag(rawValue: 0)) { (query, changeSet) in
+
+			guard let queryItem = query.queryItem else { return }
+
+			switch query.state {
+			case .contentsFromCache, .waitingForServerReply:
+				break
+			case .idle:
+				self.core?.stop(query)
+
+				let items = changeSet?.queryResult ?? []
+				guard items.first != nil else {
+					OnBackgroundQueue {
+						self.copy(item: queryItem, with: 0)
+					}
+					return
 				}
 
-				if let progress = core.copy(item, to: rootItem, withName: name, options: nil, resultHandler: { (error, _, item, _) in
-					if error != nil {
-						Log.log("Error \(String(describing: error)) duplicating \(String(describing: item?.path))")
-						self.completed(with: error)
-					} else {
-						self.completed()
+				var fileCounter = -1
+				for anItem in items {
+					guard let baseName = anItem.baseName else { return }
+
+					let pattern = "\(self.localizedCopy)[ ]?[0-9]{0,}$"
+					let matched = baseName.matches(for: pattern)
+					if matched.count > 0 {
+						let copyCounter = Int(matched.first!.replacingOccurrences(of: "\(self.localizedCopy) ", with: "")) ?? 0
+						if copyCounter > fileCounter {
+							fileCounter = copyCounter
+						}
 					}
-				}) {
-					publish(progress: progress)
 				}
+				fileCounter = (fileCounter + 1)
+				OnBackgroundQueue {
+					self.copy(item: queryItem, with: fileCounter)
+				}
+
+			case .targetRemoved: break
+
+			default: break
 			}
-		})
+		}
+	}
+
+	// MARK: - Copy Action
+	func copy(item: OCItem, with counter: Int) {
+		guard let core = self.core, let itemBaseName = item.baseName else {
+			completed(with: NSError(ocError: .itemNotFound))
+			return
+		}
+		guard let rootItem = item.parentItem(from: core) else {
+			completed(with: NSError(ocError: .itemNotFound))
+			return
+		}
+
+		let pattern = " \(localizedCopy)[ ]?[0-9]{0,}$"
+		var copyFileName = "\(itemBaseName.replacingOccurrences(for: pattern)) \(localizedCopy)"
```

Comment:
@hosy can we assume that all NOT yet duplicated files doesn't have "copy x" in the name? What if the file was named like that intentionally. So, try to duplicate a file and then duplicate a resulting duplicate in macOS.. You'll see what I mean. So if you duplicate a file `a.png`, it will create `a 2.png` and then if you duplicate `a 2.png`, it will produce `a 2 2.png`

Reply:
What I'm doing here, is exactly what you can find in macOS Finder. If you duplicate a file which was already duplicated, Finder names this also with ending` copy x`
Samples can be found in this PR **How Has This Been Tested?**

Reply:
@hosy see my another comment.. Finder behaviour seems to be different depending on how you do it)